react/features/base/tracks/actions.js

```diff
@@ -79,19 +81,64 @@ export function createLocalTracksA(options = {}) {
         // to implement them) and the right thing to do is to ask for each
         // device separately.
         for (const device of devices) {
-            createLocalTracksF(
-                {
-                    cameraDeviceId: options.cameraDeviceId,
-                    devices: [ device ],
-                    facingMode: options.facingMode || CAMERA_FACING_MODE.USER,
-                    micDeviceId: options.micDeviceId
-                },
-                /* firePermissionPromptIsShownEvent */ false,
-                store)
-            .then(
-                localTracks => dispatch(_updateLocalTracks(localTracks)),
-                reason =>
-                    dispatch(_onCreateLocalTracksRejected(reason, device)));
+            if (getState()['features/base/tracks']
+                    .find(t => t.local && t.mediaType === device)) {
+                throw new Error(`Local track for ${device} already exists`);
+            }
+
+            const gumProcess
+                = createLocalTracksF(
+                    {
+                        cameraDeviceId: options.cameraDeviceId,
+                        devices: [ device ],
+                        facingMode:
+                            options.facingMode || CAMERA_FACING_MODE.USER,
+                        micDeviceId: options.micDeviceId
+                    },
+                    /* firePermissionPromptIsShownEvent */ false,
+                    store)
+                .then(
+                    localTracks => {
+                        // Because GUM is called for 1 device (which is actually
+                        // a media type 'audio', 'video', 'screen', etc.) we
+                        // should not get more than one JitsiTrack.
+                        if (localTracks.length !== 1) {
+                            throw new Error(
+                                `Expected exactly 1 track, but was given ${
+                                    localTracks.length} tracks for device: ${
+                                    device}.`);
+                        }
+
+                        if (gumProcess.canceled) {
+                            return _disposeTracks(localTracks)
+                                    .then(
+                                        dispatch(_trackCreateCanceled(device)));
```

Comment:
@paweldomas, I don't think the `.then()` argument is correct because it will lead to the premature and unconditional dispatch of `_trackCreateCanceled`. I think the idea for it was something like `() => dispatch(_trackCreateCanceled...)`. Could you please take a look?